core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,776 @@
+{
+#ifdef CFG_WITH_VFP
+	thread_user_clear_vfp(&spc->uctx.vfp);
+#endif
+}
+
+static TEE_Result sec_part_enter_user_mode(struct sec_part_ctx *spc)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t exceptions = 0;
+	uint32_t panic_code = 0;
+	uint32_t panicked = 0;
+	uint64_t cntkctl;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cntkctl = thread_get_cntkctl_el1();
+	thread_set_cntkctl_el1(cntkctl | 0x3);
+	__thread_enter_user_mode(&spc->regs, &panicked, &panic_code);
+	thread_set_cntkctl_el1(cntkctl);
+	thread_unmask_exceptions(exceptions);
+
+	clear_vfp_state(spc);
+
+	if (panicked) {
+		abort_print_current_ta();
+		DMSG("sec_part panicked with code 0x%"PRIx32, panic_code);
+		res = TEE_ERROR_TARGET_DEAD;
+	}
+
+	return res;
+}
+
+static void init_stmm_regs(struct sec_part_ctx *spc, unsigned long a0,
+			   unsigned long a1, unsigned long sp, unsigned long pc)
+{
+	spc->regs.x[0] = a0;
+	spc->regs.x[1] = a1;
+	spc->regs.sp = sp;
+	spc->regs.pc = pc;
+}
+
+static TEE_Result alloc_and_map_sp_fobj(struct sec_part_ctx *spc, size_t sz,
+					uint32_t prot, vaddr_t *va)
+{
+	size_t num_pgs = ROUNDUP(sz, SMALL_PAGE_SIZE) / SMALL_PAGE_SIZE;
+	struct fobj *fobj = fobj_ta_mem_alloc(num_pgs);
+	struct mobj *mobj = mobj_with_fobj_alloc(fobj, NULL);
+	TEE_Result res = TEE_SUCCESS;
+
+	fobj_put(fobj);
+	if (!mobj)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = vm_map(&spc->uctx, va, num_pgs * SMALL_PAGE_SIZE,
+		     prot, 0, mobj, 0);
+	if (res)
+		mobj_put(mobj);
+
+	return res;
+}
+
+static void *zalloc(void *opaque __unused, unsigned int items,
+		    unsigned int size)
+{
+	return malloc(items * size);
+}
+
+static void zfree(void *opaque __unused, void *address)
+{
+	free(address);
+}
+
+static void uncompress_image(void *dst, size_t dst_size, void *src,
+			     size_t src_size)
+{
+	z_stream strm = {
+		.next_in = src,
+		.avail_in = src_size,
+		.next_out = dst,
+		.avail_out = dst_size,
+		.zalloc = zalloc,
+		.zfree = zfree,
+	};
+	int st = 0;
+
+	st = inflateInit(&strm);
+	if (st != Z_OK)
+		panic("inflateInit");
+	st = inflate(&strm, Z_SYNC_FLUSH);
+	if (st != Z_STREAM_END)
+		panic("inflate");
+	st = inflateEnd(&strm);
+	if (st != Z_OK)
+		panic("inflateEnd");
+}
+
+static TEE_Result load_stmm(struct sec_part_ctx *spc)
+{
+	struct secure_partition_boot_info *boot_info = NULL;
+	struct secure_partition_mp_info *mp_info = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	vaddr_t sp_addr = 0;
+	vaddr_t image_addr = 0;
+	vaddr_t heap_addr = 0;
+	vaddr_t stack_addr = 0;
+	vaddr_t sec_buf_addr = 0;
+	vaddr_t ns_comm_buf_addr = 0;
+	unsigned int sp_size;
+
+	sp_size = ROUNDUP(stmm_image_uncompressed_size, SMALL_PAGE_SIZE) +
+			stmm_stack_size + stmm_heap_size + stmm_sec_buf_size;
+	res = alloc_and_map_sp_fobj(spc, sp_size,
+				    TEE_MATTR_PRW, &sp_addr);
+	if (res)
+		return res;
+
+	res = alloc_and_map_sp_fobj(spc, stmm_ns_comm_buf_size,
+				    TEE_MATTR_URW | TEE_MATTR_PRW,
+				    &ns_comm_buf_addr);
```

Comment:
This is misleading, `ns_comm_buf_addr` is in secure memory.

Reply:
ok, I'll rename that